nvflare/app_common/tie/py_applet.py

```diff
@@ -0,0 +1,242 @@
+            t = threading.Thread(
+                target=starter.start,
+                args=(ctx,),
+                daemon=True,
+                name="applet",
+            )
+            t.start()
+            if not starter.started:
+                self.logger.error(f"cannot start applet: {starter.error}")
+                raise RuntimeError(starter.error)
+            self.starter = starter
+        else:
+            # start in a separate local process
+            # must remove the fl_context from ctx since it's not pickleable!
+            ctx.pop(Constant.APP_CTX_FL_CONTEXT)
+            self.logger.info("starting applet in another process")
+            self.process = multiprocessing.Process(
+                target=starter.start,
+                args=(ctx,),
+                daemon=True,
+                name="applet",
+            )
+            self.process.start()
+
+    def stop(self, timeout=0.0):
+        """Stop the applet
+
+        Args:
+            timeout: amount of time to wait for the applet to stop by itself. If the applet does not stop on
+                its own within this time, we'll forcefully stop it by kill.
+
+        Returns: None
+
+        """
+        if not self.runner:
+            raise RuntimeError("PyRunner is not set")
+
+        if self.in_process:
+            self.runner.stop(timeout)
+        else:
+            p = self.process
+            self.process = None
+            if p:
+                assert isinstance(p, multiprocessing.Process)
+                if p.exitcode is None:
+                    # the process is still running
+                    if timeout > 0:
+                        # wait for the applet to stop by itself
+                        start = time.time()
+                        while time.time() - start < timeout:
+                            if p.exitcode is not None:
+                                # already stopped
+                                self.logger.info(f"applet stopped (rc={p.exitcode}) after {time.time()-start} secs")
+                                return
+                            time.sleep(0.1)
+                    self.logger.info("stopped applet by killing the process")
+                    p.kill()
+
+    def is_stopped(self) -> (bool, int):
+        if not self.runner:
+            raise RuntimeError("PyRunner is not set")
+
+        if self.in_process:
+            if self.starter:
+                if self.starter.stopped:
+                    self.logger.info("starter is stopped!")
+                    return True, self.starter.exit_code
+            return self.runner.is_stopped()
+        else:
+            if self.process:
+                assert isinstance(self.process, multiprocessing.Process)
+                ec = self.process.exitcode
+                if ec is None:
+                    return False, 0
+                else:
+                    return True, ec
+            else:
+                return True, 0
```

Comment:
I am also thinking is that, is it better to have one common class with a flag "self.in_process" to control its behavior
or we could have two classes one is ThreadedPyApplet, the other ProcessPyApplet or something.

Reply:
Not really. The reason is that the "runner" is the same for both cases.